Sources/StreamChat/Utils/Database/DatabaseItemConverter.swift

```diff
@@ -0,0 +1,57 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+/// Converts database models to immutable value types by reusing existing items.
+enum DatabaseItemConverter {
+    /// Convert database models by reusing existing unchanged items.
+    /// - Parameters:
+    ///   - dtos: A list of DTOs in the NSFetchedResultsController.
+    ///   - existing: A list of existing items.
+    ///   - changes: List changes reported by the NSFetchedResultsController.
+    ///   - itemCreator: A closure which converts database models.
+    ///   - itemReuseKeyPaths: A pair of keypaths used for matching database models to existing items.
+    ///   - sorting: A list of sort values for sorting items outside of NSFetchedResultsController.
+    ///   - checkCancellation: A closure called before each invocation fo the `itemCreator`. If true, no more models are converted.
+    /// - Returns: <#description#>
+    static func convert<Item, DTO>(
+        dtos: [DTO],
+        existing: [Item],
+        changes: [ListChange<Item>]?,
+        itemCreator: (DTO) throws -> Item,
+        itemReuseKeyPaths: (item: KeyPath<Item, String>, dto: KeyPath<DTO, String>)?,
+        sorting: [SortValue<Item>],
+        checkCancellation: () -> Bool
+    ) -> [Item] where DTO: NSManagedObject {
+        let items: [Item]
+        
+        // Reuse converted items by id
+        if let itemReuseKeyPaths, !existing.isEmpty {
```

Comment:
thinking if we should feature flag it (by default enabled though), so we can easily disable it if things go wrong. Wdyt?

Reply:
I think it makes sense. I will add one.